bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py

```diff
@@ -0,0 +1,512 @@
+                    else:
+                        self.graph_nodes += '{}'.format(elem)
+                self.graph_nodes += '}'
+
+        # Make Snapshot lazy
+        # TODO: Do a proper processing of the args (user might have specified
+        # her own options object)
+        if operation.name == 'Snapshot':
+            self.graph_nodes += ', "", lazy_options'
+
+
+    def execute(self, rdf):
+        '''
+        Compiles the workflow generation code and executes it.
+
+        Args:
+            rdf (ROOT::RDF::RNode): object that represents the dataset on
+                which to execute the workflow.
+
+        Returns:
+            tuple: the first element is the list of results of the actions in
+                the C++ workflow, the second element is the list of result types
+                corresponding to those actions.
+        '''
+
+        function_id = self._compile()
+        return self._run_function(rdf, function_id)
+
+    def _compile(self):
+        '''
+        Generates the workflow code C++ file and compiles it with ACLiC
+        into a shared library. The library is also loaded as part of the
+        `TSystem::CompileMacro` call.
+
+        The name of the generated C++ file contains both a hash of its
+        code and the ID of the process that created it. This is done to
+        prevent clashes between multiple (non-sandboxed) worker processes
+        that try to write to the same file concurrently.
+
+        A class-level cache keeps track of the workflows that have been already
+        compiled to prevent unncessary recompilation (e.g. when a worker
+        process runs multiple times the same workflow).
+
+        Returns:
+            string: the id of the function to be executed. Such id is appended
+                to CppWorkflow.FUNCTION_NAME to prevent name clashes (a worker
+                process might compile and load multiple workflow functions).
+                A function id is mostly the hash of its code.
+        '''
+
+        # TODO: Make this function thread-safe? To support Dask threaded
+        # workers
+
+        code = self._get_code()
+        code_hash = hash(code)
+        # A hash can be a negative value, replace '-' with '_' so it can be
+        # part of a function name
+        function_id = str(code_hash).replace('-', '_', 1)
+        if code_hash in CppWorkflow.cached_wfs:
+            # We already compiled and loaded a workflow function with this
+            # code. Return the id of that function
+            return function_id
+
+        # We are trying to run this workflow for the first time in this
+        # process. First dump the code in a file with the right function name
+        cpp_file_name = 'rdfworkflow_{code_hash}_{pid}.cpp' \
+                        .format(code_hash=code_hash, pid=os.getpid())
+        code = code.replace(CppWorkflow.FUNCTION_NAME, CppWorkflow.FUNCTION_NAME + function_id, 1)
```

Comment:
if I understand correctly this is relying on the fact that the first occurrence of `GenerateGraph` in `code` (also as a part of longer identifiers) is at the point of the declaration. It's probably a safe bet, but maybe even safer would be to put something like `__RDF_FUNC_NAME_PLACEHOLDER__` in `code` and replace that?

Reply:
Yes ok!